### PR TITLE
feat(useExplicitFunctionReturnType): support typed function expressions

### PR DESCRIPTION
**Pull request by kaykdm**
## Summary
related: https://github.com/biomejs/biome/issues/2017

This PR introduces the ability to detect typed function expressions in the useExplicitFunctionReturnType rule.

todos:
- [x] A function with a type assertion using `as` or `<>`
```
const asTyped = (() => '') as () => string;
const castTyped = <() => string>(() => '');
```
- [x] A variable declarator with a type annotation
```
type FuncType = () => string;
const arrowFn: FuncType = () => 'test';
```
- [x] A function as a default parameter with a type annotation
```
type CallBack = () => void;
const f = (gotcha: CallBack = () => { }): void => { };
```
- [x]  A class property with a type annotation
```
type MethodType = () => void;
  class App {
     private method: MethodType = () => { };
}
```
- [x]  A function as a property or a nested property of a typed object
```
const x: Foo = { prop: () => {} }
const x = { prop: () => {} } as Foo
const x = <Foo>{ prop: () => {} }
const x: Foo = { bar: { prop: () => {} } }
```

### CODE REVIEW

File: crates/biome_js_analyze/src/lint/nursery/use_explicit_function_return_type.rs [hunk 1]
```diff
@@ -313,22 +355,29 @@ fn is_direct_const_assertion_in_arrow_functions(func: &AnyJsFunction) -> bool {
...
+/// ```typescript
+/// (function () {});
+/// (() => {})();
+/// ```
+fn is_iife(func: &AnyJsFunction) -> bool {
+    func.syntax()
+        .parent()
+        .and_then(JsParenthesizedExpression::cast)
+        .and_then(|expr| expr.parent::<JsCallExpression>())
+        .is_some()
```
[the remark this answers is not in the document]
Comment (kaykdm) [reply crates/biome_js_analyze/src/lint/nursery/use_explicit_function_return_type.rs]: I've updated the code as suggested!!

File: crates/biome_js_analyze/src/lint/nursery/use_explicit_function_return_type.rs [hunk 2]
```diff
@@ -392,3 +441,114 @@ fn is_first_statement_function_return(statements: JsStatementList) -> bool {
...
+            kind,
+            JsSyntaxKind::TS_AS_EXPRESSION | JsSyntaxKind::TS_TYPE_ASSERTION_EXPRESSION
+        )
+    }
+
+    syntax.parent().map_or(false, |parent| {
+        if parent.kind() == JsSyntaxKind::JS_PARENTHESIZED_EXPRESSION {
+            parent
+                .parent()
+                .map_or(false, |grandparent| is_assertion_kind(grandparent.kind()))
```
[the remark this answers is not in the document]
Comment (kaykdm) [reply crates/biome_js_analyze/src/lint/nursery/use_explicit_function_return_type.rs]: I've updated the code as suggested!!

[the remark this answers is not in the document]
Comment (Conaclos) [reply crates/biome_js_analyze/src/lint/nursery/use_explicit_function_return_type.rs]: The suggestion can be applied in other places.

File: crates/biome_js_analyze/src/lint/nursery/use_explicit_function_return_type.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Conaclos) [reply crates/biome_js_analyze/src/lint/nursery/use_explicit_function_return_type.rs]: The suggestion can be applied in other places.

[the remark this answers is not in the document]
Comment (kaykdm) [reply crates/biome_js_analyze/src/lint/nursery/use_explicit_function_return_type.rs]: In most other places, the argument for the functions is `SyntaxNode`. How can I shorten the function in that case?

[the remark this answers is not in the document]
Comment (Conaclos) [reply crates/biome_js_analyze/src/lint/nursery/use_explicit_function_return_type.rs]: Oh true. That's fine. Thanks!

**Review by Conaclos — APPROVED** [review]
You made it! Thanks for your amazing contribution 💗

**Pull request merged**